devdoc/constbuffer_array_requirements.md

```diff
@@ -0,0 +1,174 @@
+# constbuffer_array_requirements
+================
+
+## Overview
+
+`constbuffer_array` is a module that stiches several `CONSTBUFFER_HANDLE`s together. `constbuffer_array` can add/remove a `CONSTBUFFER_HANDLE` at the beginning (front) of the already constructed stitch. 
+
+`CONSTBUFFER_ARRAY_HANDLE`s are immutable, that is, adding/removing a `CONSTBUFFER_HANDLE` to/from an existing `CONSTBUFFER_ARRAY_HANDLE` will result in a new `CONSTBUFFER_ARRAY_HANDLE`.
+
+## Exposed API
+
+```c
+typedef struct CONSTBUFFER_ARRAY_HANDLE_DATA_TAG* CONSTBUFFER_ARRAY_HANDLE;
+
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create, const CONSTBUFFER_HANDLE*, buffers, uint32_t, buffer_count);
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_empty);
+
+MOCKABLE_FUNCTION(, void, constbuffer_array_inc_ref, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle);
+MOCKABLE_FUNCTION(, void, constbuffer_array_dec_ref, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle);
+
+/*add in front*/
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_add_front, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, CONSTBUFFER_HANDLE, constbuffer_handle);
+
+/*remove front*/
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_remove_front, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, CONSTBUFFER_HANDLE *const_buffer_handle);
+
+/* getters */
+MOCKABLE_FUNCTION(, int, constbuffer_array_get_buffer_count, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, uint32_t*, buffer_count);
+MOCKABLE_FUNCTION(, CONSTBUFFER_HANDLE, constbuffer_array_get_buffer, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, uint32_t, buffer_index);
+```
+
+### constbuffer_array_create_empty
+
+```c
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create, const CONSTBUFFER_HANDLE*, buffers, uint32_t, buffer_count);
+```
+
+`constbuffer_array_create` creates a new const buffer array made of the const buffers in `buffers`.
+
+**SRS_CONSTBUFFER_ARRAY_01_009: [** `constbuffer_array_create` shall allocate memory for a new `CONSTBUFFER_ARRAY_HANDLE` that can hold `buffer_count` buffers. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_010: [** `constbuffer_array_create` shall clone the buffers in `buffers` and store them. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_011: [** On success `constbuffer_array_create` shall return a non-NULL handle. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_012: [** If `buffers` is NULL, `constbuffer_array_create` shall fail and return NULL. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_013: [** If `buffer_count` is 0, `constbuffer_array_create` shall fail and return NULL. **]**
```

Comment:
>buffer_count [](start = 40, length = 12)

consider allowing NULL for buffers when the count is 0.

Reply:
Ok. Will do

---
In reply to: [234323603](https://github.com/Azure/azure-c-shared-utility/pull/226#discussion_r234323603) [](ancestors = 234323603)